Modules/_testcapi/unicode.c

```diff
@@ -220,109 +220,61 @@ unicode_copycharacters(PyObject *self, PyObject *args)
     return Py_BuildValue("(Nn)", to_copy, copied);
 }
 
-/* Test PyUnicode_ToLower() */
 static PyObject *
-unicode_tolower(PyObject *self, PyObject *arg)
+unicode_case_operation(PyObject *str, int (*function)(Py_UCS4, Py_UCS4 *, int), const char *name)
 {
-    if (PyUnicode_GET_LENGTH(arg) != 1) {
-        PyErr_SetString(PyExc_ValueError, "unicode_tolower only accepts 1-character strings");
+    if (PyUnicode_GET_LENGTH(str) != 1) {
+        PyErr_Format(PyExc_ValueError, "%s only accepts 1-character strings", name);
         return NULL;
     }
 
-    Py_UCS4 c = PyUnicode_READ_CHAR(arg, 0);
+    Py_UCS4 c = PyUnicode_READ_CHAR(str, 0);
 
-    Py_UCS4 lower[3];
-    int chars = PyUnicode_ToLower(c, lower, Py_ARRAY_LENGTH(lower));
-    assert(chars >= 1);
+    Py_UCS4 buf[3];
+    int chars = function(c, buf, Py_ARRAY_LENGTH(buf));
+    if (chars <= 0) {
```

Comment:
```suggestion
    if (chars < 0) {
```

Reply:
0 is also an unexpected result here that shouldn't be happening.

Reply:
0 creates an empty string and the caller (Python test) will fail since it's not the expected result.